modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityTwinningKalidindiUpdate.C

```diff
@@ -0,0 +1,339 @@
+      "propagation of twins on conplanar systems is favored at early deformation stages.");
+  params.addRangeCheckedParam<Real>(
+      "upper_limit_twin_volume_fraction",
+      0.8,
+      "upper_limit_twin_volume_fraction>0&upper_limit_twin_volume_fraction<=1",
+      "The maximumum amount of twinning volume fraction allowed");
+
+  return params;
+}
+
+CrystalPlasticityTwinningKalidindiUpdate::CrystalPlasticityTwinningKalidindiUpdate(
+    const InputParameters & parameters)
+  : CrystalPlasticityStressUpdateBase(parameters),
+
+    _total_twin_volume_fraction(declareProperty<Real>(_base_name + "total_volume_fraction_twins")),
+    _total_twin_volume_fraction_old(
+        getMaterialPropertyOld<Real>(_base_name + "total_volume_fraction_twins")),
+    _initial_total_twin_volume_fraction(getParam<Real>("initial_total_twin_volume_fraction")),
+    _twin_volume_fraction(
+        declareProperty<std::vector<Real>>(_base_name + "twin_system_volume_fraction")),
+    _twin_volume_fraction_old(
+        getMaterialPropertyOld<std::vector<Real>>(_base_name + "twin_system_volume_fraction")),
+    _twin_volume_fraction_increment(
+        declareProperty<std::vector<Real>>(_base_name + "twin_system_volume_fraction_increment")),
+    _reference_strain_rate(getParam<Real>("twin_reference_strain_rate")),
+    _rate_sensitivity_exponent(getParam<Real>("twin_strain_rate_sensitivity_exponent")),
+    _characteristic_twin_shear(getParam<Real>("characteristic_twin_shear")),
+    _twin_initial_lattice_friction(getParam<Real>("initial_twin_lattice_friction")),
+    _non_coplanar_coefficient_twin_hardening(
+        getParam<Real>("non_coplanar_coefficient_twin_hardening")),
+    _coplanar_coefficient_twin_hardening(getParam<Real>("coplanar_coefficient_twin_hardening")),
+    _noncoplanar_exponent(getParam<Real>("non_coplanar_twin_hardening_exponent")),
+    _limit_twin_volume_fraction(getParam<Real>("upper_limit_twin_volume_fraction"))
+{
+  // resize local caching vectors used for substepping
+  _previous_substep_twin_resistance.resize(_number_slip_systems);
+  _previous_substep_twin_volume_fraction.resize(_number_slip_systems);
+  _twin_resistance_before_update.resize(_number_slip_systems);
+  _twin_volume_fraction_before_update.resize(_number_slip_systems);
+}
+
+void
+CrystalPlasticityTwinningKalidindiUpdate::initQpStatefulProperties()
+{
+  CrystalPlasticityStressUpdateBase::initQpStatefulProperties();
+
+  // Resize constitutive-model specific material properties
+  _twin_volume_fraction[_qp].resize(_number_slip_systems);
+  _twin_volume_fraction_increment[_qp].resize(_number_slip_systems);
+
+  // Set constitutive-model specific initial values from parameters
+  _total_twin_volume_fraction[_qp] = _initial_total_twin_volume_fraction;
+  const Real twin_volume_fraction_per_system =
+      _initial_total_twin_volume_fraction / _number_slip_systems;
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
```

Comment:
The cool kids use
```suggestion
  for (auto i: make_range(_number_slip_systems))
```

these days (requires `#inlcude "libmesh/int_range.h"`)

Reply:
well nothing about me is cool, so I guess that means I can ignore this suggestion ;-)